DxDispatch/src/dxdispatch/Executor.cpp

```diff
@@ -158,18 +172,32 @@ void Executor::operator()(const Model::DispatchCommand& command)
             }
 
             PIXBeginEvent(m_device->GetCommandQueue(), PIX_COLOR(255, 255, 0), "Dispatch '%s'", command.dispatchableName.c_str());
-            
+
+            uint32_t timestampIndex = (iteration % timestampCount) * 2;
+
+            m_device->GetCommandList()->EndQuery(m_device->GetQueryHeap(), D3D12_QUERY_TYPE_TIMESTAMP, timestampIndex);
             dispatchable->Dispatch(command);
+            m_device->GetCommandList()->EndQuery(m_device->GetQueryHeap(), D3D12_QUERY_TYPE_TIMESTAMP, timestampIndex + 1);
```

Comment:
Unfortunately, this approach (calling EndQuery on the `Device` command list) only works for DML & HLSL dispatchables. ONNX dispatchables don't use the command list stored in the `Device` class since the DirectML EP inside ORT manages work scheduling. You'll need to handle these cases differently.

One option is to move the EndQuery calls into the implementation of each dispatchable. For example, you could modify `Dispatchable::Dispatch` to include parameters for the start & end indices for the query heap. Both the HLSL & DML dispatchables would simply call EndQuery at the start & end of their implementation of `Dispatchable::Dispatch`.

The ONNX dispatchable is a little more complicated since you don't have access to the command list used by ORT. You'll need to write timestamps using the device command list, but you can't simply call `EndQuery` alone: the device command list will need to be executed for each EndQuery independently of the work done in ORT. Furthermore, you really want to write the end timestamp after `Session::Run` but before `IOBindings::SynchronizeOutputs` (which blocks the CPU until the GPU is done). Something like this:

```cpp
m_device->GetCommandList()->EndQuery(...);
m_device->ExecuteCommandList();
m_session->Run(runOptions, *m_ioBindings);
m_device->GetCommandList()->EndQuery(...);
m_device->ExecuteCommandList();
m_ioBindings->SynchronizeOutputs();
```

There is no `Device::ExecuteCommandList` currently, and this function would need to execute the device command list without blocking the CPU and waiting work to finish. It would be similar to `DispatchAndWait` (really should be called `ExecuteCommandListAndWait`) without waiting on the fence and resetting the resources.

Reply:
I added some new functions to the Onnx dispatchable, DispatchDontWait() and Wait(). Then I have a case to go and call those in the right order. The class functions don't look that nice, I welcome any suggestions how to clean them up.

I also added a function DispatchDontWait() to the device itself, which is the ExecuteCommandList() function. Let me know if I should go through and rename these functions to the "Execute" style.

Reply:
You definitely need a new API similar to Wait(), but it could be part of the Dispatchable interface.

I think it's fine to leave this as-is for now. I'll have another PR soon that adds some PIX-related enhancements and it needs a similar refactoring. I'll address this in the PIX PR and you can let me know what you think.